### DISCUSSION TITLE
Circular Scrolling List Version 6 (Released)

### DISCUSSION THREAD

**Discussion by LanKuDot** [LanKuDot/game_modules#26] [Ideas]
Version 6 is completed! 🎉
Waiting for the unity asset store to review the package.

New features overview: https://youtu.be/6y-_MaeWIAg

## Version 6 Development Roadmap

1. ✅ Optimize the input handling
2. ✅ Modularize the box transform controlling
3. ✅ Combine all the controlling modes
4. ✅ Simplify the boxes assignment, such as prefab and object pool
5. ✅ Layout preview in the editor
6. ✅ Add new event callbacks for the list item (selected and unselected)
    - `OnFocusingBoxChanged(prevBox, curBox)` (`OnCenteredContentChanged` will be replaced)
    - `OnBoxSelected(box)` (`OnBoxClick(int)` will be replaced)
7. ✅ Add new exposed variables
    - `CircularScrollingList.listBoxes`: The list boxes belong to this list
    - `LisBox.ScrollingList`: The list which this boxes belongs to
8. ✅ An option for ending the list at two ends of the list in the linear mode
    - Modify the focusing algorithm
    - Apply to linear mode
9. ✅ Fix bugs because of new features
10. ✅ Write the documentation and make the video

**Comment by hacecc** [comment]
Hello, it would be great if you could use multiple control modes(Drag, Mouse Wheel, Function) at the same time. I noticed that if you use a duplicate copy of the CircularScrollingList script component with a different control mode it kinda works with some bugs.

**Reply by LanKuDot** [reply]
Hi, thanks for the suggestion. I will try it.
But it will be difficult for me, because these modes are using different control methods.

**Reply by baroquedub** [reply]
Came here to say the same. It's the only thing that's stopping me using this great asset

**Reply by LanKuDot** [reply]
OK, it's time for me to face the problem. 😭

**Reply by LanKuDot** [reply]
Hi. I am still developing the version 6.
But I made the quick solution for multi-controlling mode.
You can check it in https://github.com/LanKuDot/game_modules/discussions/34

**Reply by LanKuDot** [reply]
I finally completed the new version. 🎉

**Comment by Mykey05** [comment]
Hi, Any way of stopping the first/last elements from being dragged and leaving an empty gap between the rect box and list boxes?
![image](https://user-images.githubusercontent.com/39998806/158082742-978b4600-d2a1-483c-9a87-cfa30bad732e.png)

Just for context: in my list, i have 3 list boxes that fill the whole rect transform and the centered element is the 2nd one. So assuming my list only has these 3 items I would like to stop it from being dragged around.

**Reply by LanKuDot** [reply]
Hi, I think it could be done by checking the content size in the bank. If the size is smaller than 3, then inactivate the scrolling list by `enabled`.

**Reply by Mykey05** [reply]
i have found a way to fix something like that using the existing functionality.
 `public void OnListCenteredContentChanged(int centeredContentID)
	{
		currentCenterId = centeredContentID;
	}
	public void OnMovementEnd()
	{
		if (currentCenterId == 0)
		{
			list.SelectContentID(1);
		}
		if (currentCenterId == GetListLength() - 1)
		{
			list.SelectContentID(GetListLength() - 2);
		}
	}`

Attach these functions to the ciruclar list events. In my case 1st element and n-1 element will show all elements properly with no gaps in the top/bottom.

**Reply by LanKuDot** [reply]
Nice!

**Comment by Myonmu** [comment]
Hi, this might be a rather specific use case but it might be interesting to have methods that allow to reorder the list. By reordering the list, I mean by keeping the centered box centered but pop out a little bit, and when scrolling the list, it is the rest of the list that scrolls, allowing the centered box to be inserted to another place of the list.

**Reply by LanKuDot** [reply]
It sounds like rearrange the list, but it is not by dragging the target element, it is by making the whole list to fit the target element. Right?

**Reply by Myonmu** [reply]
for example:
```
[1][2][3][4][5]  (original list)

        [4] (the selected item pops out of the list a little)
[1][2][3] [5]

        [4]
   [1][2] [3][5] (the rest of the list is being moved by move functions)

        [4]
          [1][2][3][5]

        [4][1][2][3][5] (the selected item is put back to the list)
```

**Reply by LanKuDot** [reply]
Yeah. That is hard to implement. The moving mechanism is different.
Maybe it is possible once the list supports the dynamic arrangement.
